baybe/searchspace/continuous.py

```diff
@@ -205,18 +213,86 @@ def samples_random(self, n_points: int = 1) -> pd.DataFrame:
         # without the get_polytope_samples, which means torch and botorch
         # wouldn't be needed.
 
-        points = get_polytope_samples(
-            n=n_points,
-            bounds=torch.from_numpy(self.param_bounds_comp),
-            equality_constraints=[
-                c.to_botorch(self.parameters) for c in self.constraints_lin_eq
-            ],
-            inequality_constraints=[
-                c.to_botorch(self.parameters) for c in self.constraints_lin_ineq
-            ],
-        )
+        if len(self.constraints_cardinality) != 0:
+            return self._sample_with_cardinality_constraints(n_points)
+        else:
+            points = get_polytope_samples(
+                n=n_points,
+                bounds=torch.from_numpy(self.param_bounds_comp),
+                equality_constraints=[
+                    c.to_botorch(self.parameters) for c in self.constraints_lin_eq
+                ],
+                inequality_constraints=[
+                    c.to_botorch(self.parameters) for c in self.constraints_lin_ineq
+                ],
+            )
+            return pd.DataFrame(points, columns=self.param_names)
+
+    def _sample_with_cardinality_constraints(self, n_points: int = 1) -> pd.DataFrame:
+        """Get random samples from the continuous space with cardinality constraints.
+
+        Args:
+            n_points: see sample_random()
+
+        Returns:
+            see sample_random()
+        """
+        import torch
+        from botorch.utils.sampling import get_polytope_samples
+        # TODO Revisit: torch and botorch here are actually only necessary if there
+        # are constraints. If there are none and the lists are empty we can just sample
+        # without the get_polytope_samples, which means torch and botorch
+        # wouldn't be needed. see samples_random().
+
+        if len(self.constraints_cardinality) != 0:
+            # generate all possibilities of inactive parameters which are compatible
+            # to all cardinality constraints
+            _, inactive_params_full_list = get_combinations_from_cardinality_con(
+                list(self.constraints_cardinality)
+            )
 
-        return pd.DataFrame(points, columns=self.param_names)
+            points_all = torch.Tensor()
+            i_ite = 0
+            n_ite_thres = 1e5
+            while len(points_all) < n_points:
+                # randomly choose one set of inactive parameters
+                inactive_params_sampled = random.choice(inactive_params_full_list)
+                bounds = self.param_bounds_comp
+
+                # set each inactive parameter to zero by changing bounds to [0, 0]
+                if len(inactive_params_sampled) != 0:
+                    for idx, param in enumerate(self.parameters):
+                        if param.name in inactive_params_sampled:
+                            bounds[:, idx] = [0, 0]
+                try:
+                    point = get_polytope_samples(
```

Comment:
This is copy-pasted and can be avoided by cleverly reusing the code from sampling without cardinality constraints